[API-1530] Update Compact serialization documentation (#554)
hazelcast/hz-docs: docs/modules/serialization/pages/compact-serialization.adoc

separately from the classes.

It also supports zero-configuration use cases by automatically extracting schemas out of the
Java classes and records, and .NET classes which are cached and reused later, with no extra cost.

The underlying format of the Compact serialized objects is platform and language independent.

NOTE: Compact serialization is promoted to the stable status in the 5.2 Hazelcast release.

> The older versions released under the BETA status are not compatible with stable the 5.2 version. +

Author: This is to create a line break and start the next sentence in a new paragraph

separately from the classes.

It also supports zero-configuration use cases by automatically extracting schemas out of the
Java classes and records, and .NET classes which are cached and reused later, with no extra cost.

The underlying format of the Compact serialized objects is platform and language independent.

NOTE: Compact serialization is promoted to the stable status in the 5.2 Hazelcast release.
The older versions released under the BETA status are not compatible with stable the 5.2 version. +
For future versions, the backward compatibility will be maintained, just like any other
Hazelcast feature.

> 

Reviewer: @mdumandag , don't you think this a release notes or migration guide material? (because our product documentation are versioned)

Reviewer: Maybe we can reword the related entry at https://docs.hazelcast.com/hazelcast/5.2/release-notes/5-2-0#breaking-changes

Author: Yes, we can put this information in a couple of places, but my impression was that since we were beta in 5.1 and 5.0 for compact, and only promoted to the stable status in 5.2, I don't think it is a breaking change. We never gave a guarantee of future compatibility for compact beta

Author: But, this is definitely important information, so maybe we can put it in some place in the release notes

in the serializer.

Choosing a type name will associate that name with the schema and will make the
polyglot use cases, where there are multiple clients from different languages,
possible. Serializers in different languages can work on the same data, provided
that their read and write methods are compatible, and they have the same type name.

NOTE: If you evolve your class in the later versions of your application, by adding
or removing fields, you should continue using the same type name for that class.

To register an explicit serializer for a certain class, add the following configuration:

**Programmatic Configuration:**

[tabs]
====

Java - Member::
+
--
[source,java]
----
Config config = new Config();
config.getSerializationConfig()
        .getCompactSerializationConfig()
        .addSerializer(new FooSerializer());
----
--

Java - Client::
+
--
[source,java]
----
ClientConfig clientConfig = new ClientConfig();
clientConfig.getSerializationConfig()
        .getCompactSerializationConfig()
        .addSerializer(new FooSerializer());
----
--

.NET::
+
--
[source,cs]
----

----
--

Node.js::
+
--
[source,javascript]
----

> Client.newHazelcastClient(

Author: @srknzl @harunalpak I assume this is a Javascript code, not TypeScript. Is it OK for the code samples? Same for the other Node.js code snippets

Reviewer 2: IMO, using javascript in code samples is better. Using our client with typescript is just an option. We use javascript in our code samples as well. Typescript users can easily convert these examples to TS.

    }
};
----
--

Python::
+
--
[source,python]
----
class EmployeeSerializer(CompactSerializer[Employee]):
    def read(self, reader: CompactReader):
        id = reader.read_int64("id")
        name = reader.read_string("name")
        return Employee(id, name)

    def write(self, writer: CompactWriter, employee: Employee):
        writer.write_int64("id", employee.id)
        writer.write_string("name", employee.name)

    def get_type_name(self):
        return "employee"

    def get_class(self):
        return Employee
----
--

Go::
+
--
[source,go]
----

----
--

====

The last step is to register the serializer in the member or client configuration, as shown
in the <<configuration>> section.

Upon serialization, a schema will be created from the serializer, and a unique schema identifier
will be assigned to it automatically.

After the configuration registration, Hazelcast will serialize instances of the `Employee`
class using the `EmployeeSerializer`.

== Supported Types

Compact serialization supports the following list as first class types. Any other type
can be implemented on top of these, by using these types as building blocks.

[cols="1m,1a,1a,1a,1a,1a,1a,1a]
|===

> | Type | Java | .NET | C++ | Node.js | Python | Go | Description

Author: @harunalpak can you also please fill this table with the type mappings for Node.js?

Reviewer 3: Fixed.

|
| Optional[list[decimal.Decimal]]
|
| Array of decimals or null.

| TIME
| LocalTime
|
| boost::optional<hazelcast::client::local_time>
|
| Optional[datetime.time]
|
| Time consisting of hours, minutes, seconds, and nanoseconds or null.

| ARRAY_OF_TIME
| LocalTime[]
|
| boost::optional<std::vector<boost::optional<local_time>>>
|
| Optional[list[Optional[datetime.time]]]
|
| Array of times or null.

| DATE
| LocalDate
|
| boost::optional<hazelcast::client::local_date>
|
| Optional[datetime.date]
|
| Date consisting of year, month, and day of the month or null.

| ARRAY_OF_DATE
| LocalDate[]
|
| boost::optional<std::vector<boost::optional<local_date>>>
|
| Optional[list[Optional[datetime.date]]]
|
| Array of dates or null.

| TIMESTAMP
| LocalDateTime
|
| boost::optional<hazelcast::client::local_date_time>
|
| Optional[datetime.datetime]
|
| Timestamp consisting of year, month, day of the month, hour, minutes, seconds,
and nanoseconds or null.

| ARRAY_OF_TIMESTAMP
| LocalDateTime[]
|
| boost::optional<std::vector<boost::optional<hazelcast::client::local_date_time>>>
|
| Optional[list[Optional[datetime.datetime]]]
|
| Array of timestamps or null.

> | TIMESTAMP_WITH_TIMEZONE

Author: @OzanCansel I see that we didn't put anything to this and `ARRAY_OF_TIMESTAMP_WITH_TIMEZONE ` column for C++. Is there a reason for that?

know when they try to read data serialized by a new writer. Similarly, new readers might read
fields of the classes available in the data, when they try to read data serialized by an old writer.

Assume that the two versions of the following `Employee` class lives in the cluster.

[tabs]
====

Java::
+
--
[source,java]
----
public class Employee {
    private long id;
    private String name;
}
----
--

.NET::
+
--
[source,cs]
----

----
--

C++::
+
--
[source,cpp]
----
struct employee
{
    int64_t id;
    std::string name;
};
----
--

Node.js::
+
--
[source,javascript]
----
class Employee {
    id;
    name;
};
----
--

Python::
+
--
[source,python]
----
class Employee:
    def __init__(self, id: int, name: str):
        self.id = id
        self.name = name
----
--

Go::
+
--
[source,go]
----

----
--

====

[tabs]
====

Java::
+
--
[source,java]
----
public class Employee {
    private long id;
    private String name;
    private int age; // Newly added field
}
----
--

.NET::
+
--
[source,cs]
----

----
--

C++::
+
--
[source,cpp]
----
struct employee
{
    int64_t id;
    std::string name;

>     int32_t age;

Author: @OzanCansel can we please add the comments present in Java and Python code snippets here, to make them consistent? Note that, there are a few other code snippets down below, which lacks comments in C++

};
----
--

Node.js::
+
--
[source,javascript]
----
class Employee {
    id;
    name;

>     age;

Author: @harunalpak  can we please add the comments present in Java and Python code snippets here, to make them consistent? Note that, there are a few other code snippets down below, which lack comments in Node.js

Reviewer 3: I added comments.

        return Employee
----
--

Go::
+
--
[source,go]
----

----
--

====

The last step is to register the serializer in the member or client configuration, as shown
in the <<configuration>> section.

Upon serialization, a schema will be created from the serializer, and a unique schema identifier
will be assigned to it automatically.

After the configuration registration, Hazelcast will serialize instances of the `Employee`
class using the `EmployeeSerializer`.

== Supported Types

Compact serialization supports the following list as first class types. Any other type
can be implemented on top of these, by using these types as building blocks.

[cols="1m,1a,1a,1a,1a,1a,1a,1a]
|===
| Type | Java | .NET | C++ | Node.js | Python | Go | Description

| BOOLEAN
| boolean
|
| bool
| boolean
| bool
|
| True or false represented by a single bit as either 1 or 0.
Up to eight booleans are packed into a single byte.

| ARRAY_OF_BOOLEAN
| boolean[]
|
| boost::optional<std::vector<bool>>
| boolean[]
| Optional[list[bool]]
|
| Array of booleans or null. Up to eight boolean array items are packed into a single byte.

| NULLABLE_BOOLEAN
| Boolean
| 
| boost::optional<bool>

> | boolean

Author: @harunalpak I think this should be `boolean | null` for Node.js. The same applies to others as well. Can we make sure that the type mappings here are the same as our API at here https://github.com/hazelcast/hazelcast-nodejs-client/blob/master/src/serialization/compact/CompactWriter.ts?

Reviewer 3: Fixed

separately from the classes.

It also supports zero-configuration use cases by automatically extracting schemas out of the
Java classes and records, and .NET classes which are cached and reused later, with no extra cost.

The underlying format of the Compact serialized objects is platform and language independent.

NOTE: Compact serialization is promoted to the stable status in the 5.2 Hazelcast release.

> The older versions released under the BETA status are not compatible with stable the 5.2 version. +

Reviewer: ... with the stable 5.2 version.

    // reader.ReadInt32("age") would panic if the "age" field
    // does not exist in data.
    var age int32
    if reader.getFieldKind("age") == serialization.FieldKindInt32 {
        age = reader.ReadInt32("age")
    }
    return &Employee{ID: id, Name: name, Age: age}
}

----
--

====

Note that, when an old reader reads data written by an old writer, or a new reader reads a data
written by a new writer, they will be able to read all fields written.

One thing to be careful while evolving the class is to not have any conditional code
in the `write` method. That method must write all the fields available in the current version
of the class to the writer, with appropriate field names and types. Hazelcast uses the `write`
method of the serializer to extract a schema out of the object, hence any conditional code
that may or may not run depending on the object in that method might result in an undefined
behavior.

== GenericRecord Representation

Compact serialized objects can also be represented by a `GenericRecord`, without requiring
the class in the classpath. See xref:clusters:accessing-domain-objects.adoc[].

== SQL Support

> Compact serialized objects can be used in SQL statements, provided that a mappings are created,

Reviewer 2: ```suggestion
Compact serialized objects can be used in SQL statements, provided that a mapping is created,
```

Author: Changed it to `provided that mappings are created`

    }
}
----
--

.NET::
+
--

> [source,cs]

Author: Yes, I will coordinate the effort for adding these once the PR is merged.

|
| Array of decimals or null.

| TIME
| LocalTime
|
|
|
| Optional[datetime.time]
|
| Time consisting of hours, minutes, seconds, and nanoseconds or null.

| ARRAY_OF_TIME
| LocalTime[]
|
|
|
| Optional[list[Optional[datetime.time]]]
|
| Array of times or null.

| DATE
| LocalDate
|
|
|
| Optional[datetime.date]
|
| Date consisting of year, month, and day of the month or null.

| ARRAY_OF_DATE
| LocalDate[]
|
|
|
| Optional[list[Optional[datetime.date]]]
|
| Array of dates or null.

| TIMESTAMP
| LocalDateTime
|
|
|
| Optional[datetime.datetime]
|
| Timestamp consisting of year, month, day of the month, hour, minutes, seconds,
and nanoseconds or null.

| ARRAY_OF_TIMESTAMP
| LocalDateTime[]
|
|
|
| Optional[list[Optional[datetime.datetime]]]
|
| Array of timestamps or null.

| TIMESTAMP_WITH_TIMEZONE
| OffsetDateTime
|
|
|
| Optional[datetime.datetime]
|
| Timestamp with timezone consisting of year, month, day of the month, hour, minutes, seconds,
nanoseconds, and offset seconds or null.

| ARRAY_OF_TIMESTAMP_WITH_TIMEZONE
| OffsetDateTime[]
|
|
|
| Optional[list[Optional[datetime.datetime]]]
|
| Array of timestamp with timezones or null.

| COMPACT

> | Can be any type.

Author: I changed it to `Can be any user type`. Is it OK?

[tabs]
====

Java::
+
--
[source,java]
----
HazelcastInstance client = HazelcastClient.newHazelcastClient();
IMap<Long, Employee> map = client.getMap("employees");
Employee employee = new Employee(1L, "John Doe");
map.set(1L, employee);
Employee employeeFromMap = map.get(1L);
----
--

.NET::
+
--
[source,cs]
----

----
--

====

Currently, Hazelcast supports extracting schemas out of classes that have the field types shown
above and some others, for user convenience.

For Java, the zero-config serializer supports the following extra field types on top of the
first class types:

* `char`, represented as an `INT16`
* `Character`, represented as a `NULLABLE_INT16`
* Enum, represented as a `STRING`, using the names of the enum members.
* Arrays of the types listed above, represented by their respective arrays.
* `List` or `ArrayList` of the types listed above and first class types, represented by their
respective arrays with the same field name.
* `Set` or `HashSet` of the types listed above and first class types, represented by their
respective arrays with the same field name.
* `Map` or `HashMap` or the types listed above and first class types, represented by two arrays

> per field, one for keys and one for values, represented by their respective arrays for key and

Author: Makes sense. I will remove it

> per field, one for keys and one for values, represented by their respective arrays for key and

Author: I meant for each Map field of class, we will have two fields in the binary

value types. The names of those arrays are of the form `fieldName + '!keys'` and
`fieldName + '!values'`.

For Java APIs, a zero-config Compact serializer uses reflection to read and write to fields of
objects, regardless of whether those fields are public. If you use Java 9 onwards, you must
enable reflective access for packages of your module by using the `opens` statement in the
module declaration:

> [source,text]

Author: It is not following the Java syntax, and the renderer (at least in my IDE) does not colorize it if I change that to Java. So, I plan to keep it as it is

+
--
[source,xml]
----
<hazelcast>
    <serialization>
        <compact-serialization>
            <serializers>
                <serializer>
                    com.example.FooSerializer
                </serializer>
            </serializers>
        </compact-serialization>
    </serialization>
</hazelcast>
----
--

Java - Member - YAML::
+
--
[source,yaml]
----
hazelcast:
  serialization:
    compact-serialization:
      serializers:
        - serializer: com.example.FooSerializer
----
--

Java - Client - XML::
+
--
[source,xml]
----
<hazelcast-client>
    <serialization>
        <compact-serialization>
            <serializers>
                <serializer>
                    com.example.FooSerializer
                </serializer>
            </serializers>
        </compact-serialization>
    </serialization>
</hazelcast-client>
----
--

Java - Client - YAML::
+
--
[source,yaml]
----
hazelcast-client:
  serialization:
    compact-serialization:
      serializers:
        - serializer: com.example.FooSerializer
----
--

====

Lastly, the following is a sample configuration that registers zero-config
serializer for a certain class, without implementing an explicit serializer.

> This way, one can override other serialization mechanisms for a certain class
> such as Java serializable, using zero-config serializer.

Reviewer 2: ```suggestion
This way, one can override the serializer of a certain class, such as the Java `Serializable` serializer, 
with the zero-config serializer.
```

Java - Member - YAML::
+
--
[source,yaml]
----
hazelcast:
  serialization:
    compact-serialization:
      classes:
        - class: com.example.Bar
----
--

Java - Client - XML::
+
--
[source,xml]
----
<hazelcast-client>
    <serialization>
        <compact-serialization>
            <classes>
                <class>
                    com.example.Bar
                </class>
            </classes>
        </compact-serialization>
    </serialization>
</hazelcast-client>
----
--

Java - Client - YAML::
+
--
[source,yaml]
----
hazelcast-client:
  serialization:
    compact-serialization:
      classes:
        - class: com.example.Bar
----
--

====

If you want to override the serialization mechanism used for `Serializable` or
`Externalizable` classes and use Compact serialization without writing any
serializer in Java, you must add those classes to the configuration.

== GenericRecord Representation

Compact serialized objects can also be represented by a `GenericRecord`, without requiring
the class in the classpath. See xref:clusters:accessing-domain-objects.adoc[].

== SQL Support

Compact serialized objects can be used in SQL statements, provided that a mappings are created,

> similar to other serialization formats. See xref:sql:mapping-to-maps.adoc[], to learn more

Reviewer 2: maybe give a link to the `#compact-objects` section

about Compact serialization mappings.

== WAN Support

Hazelcast supports WAN replication of the Compact serialized objects between different clusters.

However, since the Compact serialization is promoted to the stable status in 5.2, and it is not
compatible with the previous BETA versions, one has to make sure that the whole WAN cluster

> topology, including all senders and receivers, is at least as new as 5.2, before start replicating data structures containing Compact

Reviewer 2: ```suggestion
topology, including all senders and receivers, are at least as new as 5.2, before starting to replicate data structures containing Compact
```
I was not going to check grammar but, am I missing something, should not this be plural ? Also, it should be `before starting to replicate`

Author: I was referring to `topology`, but changed it to `topology members` and made it plural

serialized objects.

> Since it has promoted to the stable status, it will be possible in the future releases to
> replicate Compact serialized objects between different WAN clusters of versions at least as
> new as 5.2.

Reviewer 2: ```suggestion
Since it has promoted to the stable status, replicating Compact serialized objects between WAN clusters are of different versions will be possible in the future releases provided that all clusters' versions are at least 5.2.
```

Also I think `at least as new as 5.2` feels long. If you change it, don't forget to change other occurrences.

Author: I have removed it from here, but left it in the first paragraph, as I believe it was necessary there

Java and .NET.

Compact serialization can also be used without registering a serializer in the member
or client configuration.

When Hazelcast cannot associate a class with any other serialization mechanism, instead of
throwing an exception directly, it tries to use zero-configuration Compact serialization
as a last effort.

Hazelcast tries to extract a schema out of the class. If successful, it registers the 
zero-config serializer associated with the extracted schema and uses it while serializing 
and deserializing instances of that class. If the automatic schema extraction fails,
Hazelcast throws an exception.

For example, assume that you have the same `Employee` class.

If you don't perform any kind of configuration change and use the instances of the class
directly, no exceptions will be thrown. Hazelcast will generate a schema out of the
`Employee` class the first time you try to serialize an object, cache it, and reuse it
for the subsequent serializations and deserializations.

The same holds true for the Java records. Hazelcast supports serializing and deserializing
Java records, without an extra configuration as well.

> Assuming the `Employee` class above were a Java record:

Reviewer 4: ```suggestion
Assuming the `Employee` class was a Java record:
```

map.set(1L, employee);
Employee employeeFromMap = map.get(1L);
----
--

.NET::
+
--
[source,cs]
----

----
--

====

Currently, Hazelcast supports extracting schemas out of classes that have the field types shown
above and some others, for user convenience.

For Java, the zero-config serializer supports the following extra field types on top of the
first class types:

* `char`, represented as an `INT16`
* `Character`, represented as a `NULLABLE_INT16`
* Enum, represented as a `STRING`, using the names of the enum members.
* Arrays of the types listed above, represented by their respective arrays.
* `List` or `ArrayList` of the types listed above and first class types, represented by their
respective arrays with the same field name.
* `Set` or `HashSet` of the types listed above and first class types, represented by their
respective arrays with the same field name.
* `Map` or `HashMap` or the types listed above and first class types, represented by two arrays
per field, one for keys and one for values, represented by their respective arrays for key and
value types. The names of those arrays are of the form `fieldName + '!keys'` and
`fieldName + '!values'`.

> For Java, zero-config Compact serializer uses reflection to read and write to fields of

Reviewer 4: ```suggestion
For Java, a zero-config Compact serializer uses reflection to read and write to fields of
```

that may or may not run depending on the object in that method might result in an undefined
behavior.

== Configuration

The configuration can be used to register either

- an explicit `CompactSerializer`
- a zero-config serializer for a class to override other serialization mechanisms,
for languages that provide zero-config Compact serialization.

In case of an explicit serializer, you have to supply a unique type name for the class
in the serializer.

Choosing a type name will associate that name with the schema and will make the
polyglot use cases, where there are multiple clients from different languages,
possible. Serializers in different languages can work on the same data, provided
that their read and write methods are compatible, and they have the same type name.

NOTE: If you evolve your class in the later versions of your application, by adding
or removing fields, you should continue using the same type name for that class.

When a class is serialized using the zero-config Compact serializer, Hazelcast will choose the
fully qualified class name for Java as the type name automatically.

> Below is the way to register an explicit serializer for a certain class.

Reviewer 4: ```suggestion
To register an explicit serializer for a certain class, add the following configuration:
```
